apache/cloudstack
# API-call to declare host as Degraded

## Description

Sometimes a host is down (admins know) but CS keeps VMs in the 'Running' state to avoid data corruption. An example of that is when a KVM host using Ceph halts. CloudStack cannot check if VM still writing on disk, therefore it leaves it in the "Running" state and HA will not be triggered for those VMs.

With that in mind, this PR proposes an API-call to declare a Host as "Degraded". All "Running"  VMs on that Host should then be set to 'Stopped' so HA can kick in.

1. Command name: `declareHostAsDegraded`; parameters: id (id of the host); description: _Declare host as 'Degraded'. The host must be on 'Disconnected' or 'Alert' state._
2. Command name: `cancelHostAsDegraded`; parameters: id (id of the host); description: _Cancel host status from 'Degraded'. The host will transit back to status 'Enabled'._

### Note:

Admin **MUST** check if there are VMs running on the problematic host and call such API command just if he/she is **100%** sure that can deploy the Running VMs on another host.

## Types of changes


## Screenshots (if appropriate):

Example of how host Resource State is shown on UI when a host is declared 'Degraded' (taken when this PR was using state "Dead"):
![image](https://user-images.githubusercontent.com/5025148/84510874-10813400-ac9c-11ea-8bff-580dc2c4d9a5.png)

## How Has This Been Tested?




**Note:** tests were performed on two KVM hosts with HA enabled: _Host01_ and _Host02_.

#### Test 01: declareHostAsDegraded
**ha.force = false**
1. Deploy VMs on host Host01
- deploy 2 VMs shared storage, HA enabled (VM01, VM02)
- deploy 1 VM shared storage, HA disabled (VM03)
- deploy 1 VM local storage, HA disabled (VM04)
2. Stop cloudstack-agent.service on Host1
3. Run Virsh command to shutdown/destroy all VMs in Host10
4. All VMs are in "Running" state, the host is in "Disconnected"
5. call API command declareHostsAsDegraded
- Host state migrated to "Degraded"
- VMs with HA + Shared started on Host02 (transit from Running to Stopped, Stopped -> Starting, Starting -> Running)
- VMs with no HA were marked as Stopped (transit from Running to Stopped, restart Scheduler ignores non-HA VMs)

**ha.force=true**
- Steps 1, 2, 3, 4, and 5 executed again.
- VMs VM01, VM02, and VM03 got re-started by HA
- VM04 did not start due to its local storage (this is the correct behavior).

Test conclusion: all working as expected.

#### Test 02:  cancelHostAsDegraded:
1. Start cloudstack-agent.service on Host01
2. Call API command cancelHostAsDegraded
3. Assert that VMs can be deployed/migrated on Host01
Test conclusion: all working as expected.
